### PR TITLE
Implemented ZC_BAN_LIST3 packet

### PR DESCRIPTION
**Pull request by AoShinRO**
* **Addressed Issue(s)**: 

* **Server Mode**: 

* **Description of Pull Request**:

### CODE REVIEW

File: src/char/int_guild.cpp [hunk 1]
```diff
@@ -308,8 +308,8 @@ int inter_guild_tosql( mmo_guild &g, int flag ){
 
 				Sql_EscapeStringLen(sql_handle, esc_name, e->name, strnlen(e->name, NAME_LENGTH));
 				Sql_EscapeStringLen(sql_handle, esc_mes, e->mes, strnlen(e->mes, sizeof(e->mes)));
-				if( SQL_ERROR == Sql_Query(sql_handle, "REPLACE INTO `%s` (`guild_id`,`account_id`,`name`,`mes`) "
-					"VALUES ('%d','%d','%s','%s')", schema_config.guild_expulsion_db, g.guild_id, e->account_id, esc_name, esc_mes) )
+				if( SQL_ERROR == Sql_Query(sql_handle, "REPLACE INTO `%s` (`guild_id`,`account_id`,`name`,`mes`,`char_id`) "
+					"VALUES ('%d','%d','%s','%s','%d')", schema_config.guild_expulsion_db, g.guild_id, e->account_id, esc_name, esc_mes, e->char_id) )
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:312]:
```suggestion
					"VALUES ('%u','%u','%s','%s','%u')", schema_config.guild_expulsion_db, g.guild_id, e->account_id, esc_name, esc_mes, e->char_id) )
```

File: sql-files/upgrades/upgrade_20240912.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `guild_expulsion` ADD COLUMN `char_id` int(11) unsigned NOT NULL default '0';
```
Comment (Lemongrass3110) [inline sql-files/upgrades/upgrade_20240912.sql:1]:
```suggestion
ALTER TABLE `guild_expulsion` ADD COLUMN `char_id` int(11) unsigned NOT NULL default '0';

```

File: src/char/int_guild.cpp [hunk 3]
```diff
@@ -499,6 +499,7 @@ std::shared_ptr<CharGuild> inter_guild_fromsql( int32 guild_id ){
 		Sql_GetData(sql_handle, 0, &data, nullptr); e->account_id = atoi(data);
 		Sql_GetData(sql_handle, 1, &data, &len); memcpy(e->name, data, zmin(len, NAME_LENGTH));
 		Sql_GetData(sql_handle, 2, &data, &len); memcpy(e->mes, data, zmin(len, sizeof(e->mes)));
+		Sql_GetData(sql_handle, 3, &data, nullptr); e->char_id = atoi(data);
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:502]:
```suggestion
		Sql_GetData(sql_handle, 3, &data, nullptr); e->char_id = strtoul(data, nullptr, 10);
```
See #1609

File: src/map/clif.cpp [hunk 4]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
 }
 
 
-/// Guild expulsion list (ZC_BAN_LIST).
-/// 0163 <packet len>.W { <char name>.24B <account name>.24B <reason>.40B }*
+/// Guild expulsion list 
+/// 0163 <packet len>.W { <char name>.24B <account name>.24B <reason>.40B }* (ZC_BAN_LIST).
 /// 0163 <packet len>.W { <char name>.24B <reason>.40B }* (PACKETVER >= 20100803)
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9101]:
```suggestion
/// 0163 <packet len>.W { <char name>.24B <reason>.40B }* (PACKETVER >= 20100803) (ZC_BAN_LIST)
```

File: src/map/clif.cpp [hunk 5]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
 }
 
 
-/// Guild expulsion list (ZC_BAN_LIST).
-/// 0163 <packet len>.W { <char name>.24B <account name>.24B <reason>.40B }*
+/// Guild expulsion list 
+/// 0163 <packet len>.W { <char name>.24B <account name>.24B <reason>.40B }* (ZC_BAN_LIST).
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9100]:
```suggestion
/// 0163 <packet len>.W { <char name>.24B <account name>.24B <reason>.40B }* (ZC_BAN_LIST)
```

File: src/map/clif.cpp [hunk 6]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
 /// 0163 <packet len>.W { <char name>.24B <reason>.40B }* (PACKETVER >= 20100803)
-void clif_guild_expulsionlist(map_session_data* sd)
-{
-#if PACKETVER < 20100803
-	const int offset = NAME_LENGTH*2+40;
-#else
-	const int offset = NAME_LENGTH+40;
-#endif
-	int fd, i, c = 0;
+/// 0a87 <packet len>.W { <charid>.L <reason>.40B }* (PACKETVER >= 20161019) (ZC_BAN_LIST2).
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9102]:
```suggestion
/// 0a87 <packet len>.W { <charid>.L <reason>.40B }* (PACKETVER >= 20161019) (ZC_BAN_LIST2)
```

File: src/map/clif.cpp [hunk 7]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
+/// 0b7c <packet len>.W { <charid>.L <reason>.40B <char name>.24B }* (PACKETVER >= 20200902) (ZC_BAN_LIST3).
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9103]:
```suggestion
/// 0b7c <packet len>.W { <charid>.L <reason>.40B <char name>.24B }* (PACKETVER >= 20200902) (ZC_BAN_LIST3)
```

File: src/map/clif.cpp [hunk 8]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
+static void clif_guild_expulsionlist(map_session_data& sd){
 
-	nullpo_retv(sd);
+	int i, c = 0;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9107]:
```suggestion
```

File: src/map/clif.cpp [hunk 9]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
+	PACKET_ZC_BAN_LIST* p = reinterpret_cast<PACKET_ZC_BAN_LIST*>( packet_buffer );
 
-	WFIFOHEAD(fd,4 + MAX_GUILDEXPULSION * offset);
-	WFIFOW(fd,0) = 0x163;
+	p->packetType = HEADER_ZC_BAN_LIST;
+	p->packetLen = sizeof(*p);
 
-	for( i = 0; i < MAX_GUILDEXPULSION; i++ )
-	{
+	for( i = 0; i < MAX_GUILDEXPULSION; i++ ){
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9117]:
```suggestion
	for( size_t i = 0, c = 0; i < MAX_GUILDEXPULSION; i++ ){
```

File: src/map/clif.cpp [hunk 10]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
 		struct guild_expulsion* e = &g->guild.expulsion[i];
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9118]:
```suggestion
		struct guild_expulsion& e = &g->guild.expulsion[i];
```

File: src/map/clif.cpp [hunk 11]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
+		if( e->account_id > 0 ){
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9119]:
```suggestion
		if( e.account_id == 0 ){
			continue;
		}

```

File: src/map/clif.cpp [hunk 12]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
+			PACKET_ZC_BAN_LIST_sub& banned = p->chars[c];
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9120]:
```suggestion
		PACKET_ZC_BAN_LIST_sub& banned = p->chars[c];
```

File: src/map/clif.cpp [hunk 13]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
-		if( e->account_id > 0 )
-		{
-			safestrncpy(WFIFOCP(fd,4 + c*offset), e->name, NAME_LENGTH);
-#if PACKETVER < 20100803
-			memset(WFIFOP(fd,4 + c*offset+24), 0, NAME_LENGTH); // account name (not used for security reasons)
-			memcpy(WFIFOP(fd,4 + c*offset+48), e->mes, 40);
-#else
-			memcpy(WFIFOP(fd,4 + c*offset+24), e->mes, 40);
+#if PACKETVER >= 20161019
+			banned.char_id = static_cast<decltype(banned.char_id)>(e->char_id);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9123]:
```suggestion
		banned.char_id = e->char_id;
```

File: src/map/clif.cpp [hunk 14]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
+#elif PACKETVER < 20100803
+			memset(banned.account_name, 0, sizeof(banned.account_name)); // account name (not used for security reasons)
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9125]:
```suggestion
		// account name (not used for security reasons)
		safestrncpy(banned.account_name, "", sizeof(banned.account_name)); 
```

File: src/map/clif.cpp [hunk 15]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
+#endif
+
+#if PACKETVER >= 20200902 || PACKETVER < 20161019
+			safestrncpy(banned.char_name, e->name, sizeof(banned.char_name));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9129]:
```suggestion
		safestrncpy(banned.char_name, e.name, sizeof(banned.char_name));
```

File: src/map/clif.cpp [hunk 16]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
 #endif
+
+			memcpy(banned.message, e->mes, sizeof(banned.message));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9132]:
```suggestion
		safestrncpy(banned.message, e->mes, sizeof(banned.message));
```

File: src/map/clif.cpp [hunk 17]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
+			p->packetLen += sizeof(banned);
 			c++;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9134]:
```suggestion
		p->packetLen += static_cast<decltype(p->packetLen)>(sizeof(banned));
		c++;
```

File: src/map/clif.cpp [hunk 18]
```diff
@@ -9096,47 +9096,46 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
 		}
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9135]:
```suggestion
```

**Review by Lemongrass3110 — APPROVED** [review]

File: src/char/int_guild.cpp [hunk 19]
```diff
@@ -499,6 +499,7 @@ std::shared_ptr<CharGuild> inter_guild_fromsql( int32 guild_id ){
 		Sql_GetData(sql_handle, 0, &data, nullptr); e->account_id = atoi(data);
```
Comment (AoShinRO) [inline src/char/int_guild.cpp:499]: @Lemongrass3110 account_id can be changed to strtoul(data, nullptr, 10); too?

File: src/map/clif.cpp [hunk 20]
```diff
@@ -9096,47 +9096,47 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
-	const int offset = NAME_LENGTH+40;
-#endif
-	int fd, i, c = 0;
-
-	nullpo_retv(sd);
+/// Guild expulsion list 
+/// 0163 <packet len>.W { <char name>.24B <account name>.24B <reason>.40B }* (ZC_BAN_LIST)
+/// 0163 <packet len>.W { <char name>.24B <reason>.40B }* (PACKETVER >= 20100803) (ZC_BAN_LIST)
+/// 0a87 <packet len>.W { <charid>.L <reason>.40B }* (PACKETVER >= 20161019) (ZC_BAN_LIST2)
+/// 0b7c <packet len>.W { <charid>.L <reason>.40B <char name>.24B }* (PACKETVER >= 20200902) (ZC_BAN_LIST3)
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9103]:
```suggestion
/// 0a87 <packet len>.W { <charid>.L <reason>.40B }* (ZC_BAN_LIST2)
/// 0b7c <packet len>.W { <charid>.L <reason>.40B <char name>.24B }* (ZC_BAN_LIST3)
```
The packetver is only necessary if the structure of the packet changed, but the packet id is still the same.

File: src/map/clif.cpp [hunk 21]
```diff
@@ -9096,47 +9096,47 @@ void clif_guild_expulsion( map_session_data& sd, const char* name, uint32 char_i
...
-	WFIFOHEAD(fd,4 + MAX_GUILDEXPULSION * offset);
-	WFIFOW(fd,0) = 0x163;
+	p->packetType = HEADER_ZC_BAN_LIST;
+	p->packetLen = sizeof(*p);
 
-	for( i = 0; i < MAX_GUILDEXPULSION; i++ )
-	{
-		struct guild_expulsion* e = &g->guild.expulsion[i];
+	for( size_t i = 0, c = 0; i < MAX_GUILDEXPULSION; i++ ){
+		struct guild_expulsion& e = g->guild.expulsion[i];
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:9116]:
```suggestion
		struct guild_expulsion& e = g->guild.expulsion[i];

```

File: src/char/int_guild.cpp (hunk 19 above)
Comment (Lemongrass3110) [reply src/char/int_guild.cpp:499]: No lets keep it like this right now, so that I can merge this PR.

**Review by Lemongrass3110 — APPROVED** [review]

**Review by aleos89 — APPROVED** [review]

**Pull request merged**